### PR TITLE
Expand QR Decomposition capabilities

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

The `QRdecomposition` has not implemented

- Decomposition on fat matrices
- Decomposition on rank-deficient matrices
- Decomposition on zero-rank matrices

In this patch, it would be able to handle cases sympy previously could not handle, as

- Fat matrices will be decomposed into square Q, and R with same dimensions as original.
- Previously, the computation was aborted when `norm` evaluated into 0, but now, it will move on to the next column.
- Zero-rank matrices will return Q forced with standard basis vectors, with R containing only zeros

Also, some other improvements are done as

- Deleted `rref` subroutine, and implemented rank check in the middle of computation

This implementation will be able to extend the capability of Classical Gram-Schmidt procedure.
Though I'm not exactly sure about the mathematical properties involved.

#### Release Notes

- matrices
  - Modified `QRdecomposition` to handle rank-deficient matrices and matrices with more columns than rows.

### CODE REVIEW

**Comment by sylee957** [comment]
I think there is some critical bugs for matrices like [[0,0,0],[0,0,0],[0,0,0],[1,1,1],[0,1,0],[0,0,0]] or such, at the moment now

**Comment by sylee957** [comment]
I have just changed the strategy, instead of doing limiting the computational boundaries (which was invalid), for fat matrices, padding the rows with zeros to make it fit into square would be a valid approach.
In this way, the original gram-schmitt doesn't have to be modified heavily.

File: sympy/matrices/matrices.py [hunk 1]
```diff
@@ -4053,34 +4053,49 @@ def QRdecomposition(self):
...
-        rank = n
-        row_reduced = mat.rref()[0]
-        for i in range(row_reduced.rows):
-            if row_reduced.row(i).norm() == 0:
-                rank -= 1
-        if not rank == mat.cols:
-            raise MatrixError("The rank of the matrix must match the columns")
+        ranked = list()
+
+        # If matrix is fat, pad additional zeros to make it square.
```
Comment (smichr) [inline sympy/matrices/matrices.py]: suggest:

# pad with additional rows to make wide matrices square;
# nOrig keeps track of original size so zeros can be trimmed from Q

File: sympy/matrices/tests/test_matrices.py [hunk 2]
```diff
@@ -813,6 +814,101 @@ def test_QR_non_square():
     assert R.is_upper
     assert A == Q*R
 
+    A = Matrix(2, 1, [1, 2])
+    Q, R = A.QRdecomposition()
+    assert Q.T * Q == eye(Q.cols)
+    assert R.is_upper
+    assert A == Q*R
+
+    # Short and fat matrices
```
Comment (smichr) [inline sympy/matrices/tests/test_matrices.py]: # wide (cols > rows) matrices

File: sympy/matrices/tests/test_matrices.py [hunk 3]
```diff
@@ -801,6 +801,7 @@ def test_QR():
 
 
 def test_QR_non_square():
+    # Tall and thin matrices
```
Comment (smichr) [inline sympy/matrices/tests/test_matrices.py]: # narrow (cols < rows) matrices

**Comment by smichr** [comment]
Just a few minor grammatical/usage suggestions.

**Comment by smichr** [comment]
I didn't mean to make it look like 
# shouting
I just forgot to put tics around the 
`# comment lines`.

**Pull request merged**